src/library/scanner/libraryscanner.cpp

```diff
@@ -452,6 +482,66 @@ void LibraryScanner::addNewTrack(TrackPointer pTrack) {
         emit(trackAdded(pTrack));
         emit(progressLoading(pTrack->getLocation()));
     } else {
-        qWarning() << "Track ("+pTrack->getLocation()+") could not be added";
+        qWarning()
+                << "Track (" + pTrack->getLocation() + ") could not be added";
+    }
+}
+
+bool LibraryScanner::changeScannerState(ScannerState newState) {
+    // Allowed State transitions:
+    // IDLE -> STARTING
+    // STARTING -> IDLE
+    // STARTING -> SCANNING
+    // SCANNING -> FINISHED
+    // FINISHED -> IDLE
+    // every state can change to CANCELING
+    // CANCELING -> IDLE
+    switch (newState) {
+    case IDLE:
+        // we are leaving STARTING  or CANCELING state
+        // m_state is already IDLE if a scan was canceled
+        m_state = IDLE;
+        m_stateMutex.unlock();
+        return true;
+    case STARTING:
+        // we need to lock the mutex during the STARTING state
+        // to prevent loosing cancel commands or start the scanner
+        // twice
+        if (m_stateMutex.tryLock()) {
+            if (m_state != IDLE) {
+                qDebug() << "LibraryScanner: Scan already in progress.";
+                m_stateMutex.unlock();
+                return false;
+            }
+            m_state = STARTING;
+            return true;
+        } else {
+            qDebug() << "LibraryScanner: mutex locked, state =" << m_state;
+            return false;
+        }
+    case SCANNING:
+        DEBUG_ASSERT(m_state == STARTING);
```

Comment:
I'm with Daniel regarding the assert. We could safely omit them. The asserts are just there for documentation purposes with the bonus of runtime checks during development/testing. Please remember to only "assert for the impossible". All possible errors have to be handled explicitly. But if one of these asserts ever fails we have reached an undefined program state resulting in undefined behavior.